Add support for Slurm Accounting

jdeamicis opened the pull request:
### Description of changes
* Add basic cluster configuration parameters to enable Slurm Accounting.
* Add validators and unit tests for validators for Slurm Accounting.
* Propagate Slurm Accounting cluster parameters to the dna.json of the head node.
* Add GetSecretValue IAM permission in case Slurm database is enabled.

### Tests
* Added unit tests for the new feature and modified some existing ones (e.g. test_all_validators).
* Ran unit tests with the patched code.
* Manual creation of cluster with Slurm accounting.
* Manual update of cluster of cluster to enable/disable the functionality.
* Integration tests to be added at a later stage.

### References
* https://github.com/aws/aws-parallelcluster-cookbook/pull/1530

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests ~~or integration tests~~** to cover the new/modified code.


Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jdeamicis committed:
Add CLI parameters for Slurm Database integration

Add config, schema and validators.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add unit tests for Database validators

Add unit tests for the DatabaseUriValidator.

Improve DatabaseUriValidator.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Expose Slurm Accounting database config parameters in head node dna.json

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Change validator error message when Database/Uri begins with slash

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Reformat Slurm Accounting CLI developments to pass linters tests

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Configure Slurm Accounting at cookbook level

Add Slurm configuration parameters to set up accounting (in slurm.conf and slurmdbd.conf)

Define new user to run the slurmdbd daemon and assign it the slurmdbd.conf files

Create script to update database password from Secrets Manager

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add GetSecretValue IAM permissions for Slurm Accounting

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>
Configure Slurm Accounting at cookbook level

Add Slurm configuration parameters to set up accounting (in slurm.conf and slurmdbd.conf)

Define new user to run the slurmdbd daemon and assign it the slurmdbd.conf files

Create script to update database password from Secrets Manager

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Fix Slurm database config object definition

Also add some small linter fixes

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Refactor Chef variables for Slurm Accounting

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Improve unit tests for Slurm Accounting

Add unit tests for the configuration of the dna.json file for Slurm Accounting

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Adapt urlparse logic to changes introduced in Python 3.9

The previous logic was based on the behavior of urlparse in Python 3.8. For more info,
see https://github.com/urllib3/urllib3/issues/2523 .

Modify changelog for the main package

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Remove useless pass statement in dummy_aws_api.py

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Add unit test case to increase code coverage

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Refactor cdk_builder_utils.py to address code-linters issue

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

EddyMM commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,94 @@
+        if not self._check_port(uri_parse):
+            return
+
+    def _check_trailing_slash(self, uri: str) -> bool:
```
The term **trailing** applies to the end of a string. I believe you meant `check_leading_slash`?

jdeamicis replied:
You are right. Fixing in the next commit.

EddyMM commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,94 @@
+            return
+
+    def _check_trailing_slash(self, uri: str) -> bool:
+        if uri and uri[0] == "/":
```
[startswith](https://docs.python.org/3/library/stdtypes.html#str.startswith) can also work.

jdeamicis replied:
Fixing in the next commit.

EddyMM commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,94 @@
+
+        # Throw error if the URI contains a scheme
+        if not self._check_scheme(uri_parse):
+            return
```
For consistency in validator behaviour, should we avoid returning on the first encounter of an error?
I believe the current behaviour is to build a list of validation failures then print all of them.
Using this `return` approach will break that pattern.

jdeamicis replied:
That's a very good question. When looking at the validators, most validation functions had a structure like `if ... elif ...`, meaning that they implicitly return only one type of validation error.

Some validators though show the pattern you are talking about (for example [`ExistingFsxNetworkingValidator`](https://github.com/aws/aws-parallelcluster/blob/874f95fac31731eae1b5156c5ca274c482d58cb4/cli/src/pcluster/validators/cluster_validators.py#L445)). But I believe there we are talking about looping through different entities to be validated (essentially one check per file system). Do you have any example about multiple validation errors for the same validated entity?

I'm not sure what the best user experience is. Should we tell the user all the criteria the string does not comply with, even if the string will most likely change to cope with the first validation error, potentially leading to a different outcome at the next iteration of the validation?

EddyMM replied:
> most validation functions had a structure like if ... elif ..., meaning that they implicitly return only one type of validation error.

This seems to happen when comparing a specific property as opposed to a range of properties.

For the intended experience, it may be worth discussing it with the wider team but it's not a technical blocker.

jdeamicis replied:
We will leave it like this for the time being. Waiting for feedback from beta testers.

jdeamicis committed:
Address review comments from https://github.com/aws/aws-parallelcluster/pull/4356

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

demartinofra commented on cli/src/pcluster/schemas/cluster_schema.py:
```diff
@@ -1329,6 +1330,26 @@ def make_resource(self, data, **kwargs):
+    )
+    # We don't want to take ownership of the whole structure of the StorageParameters config
+    # parameter in slurmdbd.conf, so we define a generic dictionary rather than a subschema.
+    slurmdbd_storage_parameters = fields.Dict(metadata={"update_policy": UpdatePolicy.COMPUTE_FLEET_STOP})
```
Leaving a note so that we can discuss whether we want to leave this as is, extend to allow any slurmdbd param to ba passed or drop for now and expect users to modify slurmdbd.conf in this first iteration.

jdeamicis replied:
Yes, we should discuss this.

jdeamicis replied:
This is going to be removed from the beta. Resolving this conversation.

demartinofra commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -122,6 +122,23 @@ def get_slurm_specific_dna_json_for_head_node(config: SlurmClusterConfig, schedu
     }
 
 
+def get_slurm_database_dna_json_for_head_node(config: BaseClusterConfig) -> dict:
```
why we need this to explicitly go into the dna.json and can't only rely on the config?

jdeamicis replied:
At first I thought the dna.json was the right way to export parameters to make them available to Chef (because it's ubiquitous in the code). After your comment I dug a bit more in the cookbook and found [this](https://github.com/aws/aws-parallelcluster-cookbook/blob/da0302f6a491987dc35b078a3a9e3f0a89d61f4a/cookbooks/aws-parallelcluster-config/recipes/config.rb#L18-L22):

```chef
include_recipe 'aws-parallelcluster-config::base'

fetch_config 'Fetch and load cluster configs' unless node['cluster']['scheduler'] == 'awsbatch'

include_recipe 'aws-parallelcluster-slurm::config' if node['cluster']['scheduler'] == 'slurm'
```

which I guess is why all the operations performed in `base.rb` currently rely on parameters defined in either `defaults.rb` or exported to the `dna.json` (e.g. directory service).

Since the configuration of Slurm happens after the config fetching operation I can now remove what I had exported into the `dna.json` and use the imported structure. Is there a reason we fetch the config only after the `base.rb` recipe has been called?

jdeamicis replied:
Actually, we do need at least a trigger (`['slurm']['database']['enabled']) for the Cloudwatch agent: https://github.com/aws/aws-parallelcluster-cookbook/pull/1532/ .

@demartinofra @davprat maybe we should just remove the `feature_conditions` for the export of `/var/log/slurmdbd.log`?

jdeamicis replied:
Done.

jdeamicis replied:
Fdm says it's OK to remove the feature_conditions.

jdeamicis replied:
Done.

demartinofra commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,128 @@
+class DatabaseUriValidator(Validator):
+    """Slurm database URI validator."""
+
+    def _validate(self, uri: str):
```
if you want to avoid all these "if not -> return" you could put in and all the function calls, the first returning False should stop the flow.

jdeamicis replied:
Done.

demartinofra commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,128 @@
+            return False
+        if not port:
+            self._add_failure(
+                f"No port specified in the URI. Assuming the use of port {default_mysql_port}",
+                FailureLevel.WARNING,
```
can this be an INFO?

jdeamicis replied:
Yes, I will change it.

jdeamicis replied:
Done

demartinofra commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,128 @@
+            )
+        for key, value in slurmdbd_storage_parameters.items():
+            if key not in SLURMDBD_STORAGE_PARAMETERS:
+                self._add_failure(
+                    f"'{key}' not available as Slurmdbd StorageParameter. The provided parameter will be ignored.",
+                    FailureLevel.INFO,
+                )
```
why not failing? a typo might go unnoticed otherwise

jdeamicis replied:
We need to discuss about this implementation. I did not create a schema for the `StorageParameter` parameter because I think we should not take ownership of that parameter (and in the longer run we should have a more generalized solution to allow users to customize the Slurm configuration).

Regarding the FailureLevel, I guess I can increase it to ERROR.

jdeamicis replied:
Done.

jdeamicis replied:
This is going to be removed from the beta anyway.

demartinofra commented on cli/src/pcluster/validators/database_validators.py:
```diff
@@ -0,0 +1,128 @@
+        return True
+
+
+class SlurmdbdStorageParametersValidator(Validator):
```
todo: review the validation/messaging flow

jdeamicis replied:
Removed from the beta.

jdeamicis committed:
Address first batch of reviews

Remove useless Slurm Database info from dna.json (it is
already available in the cluster config).

Improve readability of Database validator functions and change the
FailureLevel of some messages.

Adapt unit tests accordingly

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Address second batch of reviews

Create slurmdbd.conf only if Slurm accounting is enabled via the
ParallelCluster configuration file

Dramatically increase timeout for the "wait for slurm database" action
to cover cases when slurmdbd may try to reformat a database created
with a previous version of Slurm

Disable creation of the home directory for the slurmdbd user

Remove some validation functions from pcluster_slurm_config_generator.py
(to keep the validation only on the CLI side)

Removed all Slurm database-related info from dna.json (we now rely on the
cluster configuration file).

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

jdeamicis committed:
Merge branch 'develop' into wip/devel/slurm_accounting_cli_params

Q-Hilbtec commented:
Dear @davprat, thanks for all those changes.. Would you guys consider also enabling `jobacct_gather` (e.g. see https://aws.amazon.com/blogs/compute/enabling-job-accounting-for-hpc-with-aws-parallelcluster-and-amazon-rds/) in order to get real-time job stats via `sstat` and whatnot? I would make a separate issue but not sure of the visibility. It seems quite trivial (few lines of slurm config) to enable it now that you're getting accounting / a database as a whole.
